### PR TITLE
Richer Python interface to CoreNLP

### PR DESCRIPTION
**Pull request by kelvinguu**
Changes:
- `Client` has been expanded to return annotations from the CoreNLP server in three formats (protobuf, JSON and a new `AnnotatedDocument` object). The default `annotate` method returns `AnnotatedDocument`, which is the most structured form.
- `AnnotatedDocument` wraps the raw Document protobuf object, providing methods that are meant to mirror the [Simple CoreNLP](http://stanfordnlp.github.io/CoreNLP/simple.html) style.
- Added base classes `Document` and `Sentence`. These don't enforce much, but simply define a type for future kinds of `Document`s and `Sentence`s to extend. For example, we'd like to eventually have `LazySentence`, which would have the same methods as `AnnotatedSentence`, but only perform the requested annotations when asked.

A quick question: maybe we should simplify the package structure here a bit?
Instead of `stanza.corenlp.client.Client`, maybe we could just have `stanza.nlp.CoreNLPClient`?

(Note: the base of this pull request is not `develop` as you would expect, but a dummy branch called `old_corenlp`. That's because I already over-eagerly committed some of the changes I'm proposing to the `develop` branch. The dummy base branch just makes it easier to see all my proposed changes.)

### CODE REVIEW

**Comment by vzhong** [comment]
Awesome! This is great @kelvinguu! I am ok with simplifying the file structure.

File: stanza/corenlp/client.py [hunk 1]
```diff
@@ -16,24 +17,216 @@ def __init__(self, server='http://localhost:9000'):
...
+  def tokens(self):
+    return self._tokens
+
+  @property
+  def text(self):
+    return self.pb.text
+
+  @property
+  def pos_tags(self):
+    raise NotImplementedError
```
Comment (vzhong) [inline stanza/corenlp/client.py]: Why do these always raise errors? Can do we a lazy check for whether the field exists and then raise if the field doesn't exist?

Also I think you want to raise and instance and not a class:

raise NotImplementedError()

File: stanza/corenlp/client.py [hunk 2]
```diff
@@ -16,24 +17,216 @@ def __init__(self, server='http://localhost:9000'):
...
+# In particular, all the methods that take arguments.
+
+
+class AnnotatedSentence(Sentence):
+  def __init__(self, sentence_pb):
+    self.pb = sentence_pb
+    self._tokens = [AnnotatedToken(tok) for tok in self.pb.token]
+
+  def __getitem__(self, i):
+    return self.pb.token[i].word
```
Comment (futurulus) [inline stanza/corenlp/client.py]: Returning the token from `__getitem__` seems much more useful than returning just the word. This also means iterating over the sentence returns tokens, so I could do something like

```
>>> doc = client.annotate('The quick brown fox jumped over the lazy dog.', ['tokenize', 'ssplit', 'pos'])
>>> for sent in doc:
...     for tok in sent:
...         print('{}:{}'.format(tok.word, tok.pos))
The:DT
quick:JJ
brown:JJ
fox:NN
jumped:VBD
over:IN
the:DT
lazy:JJ
dog:NN
.:.
```

File: stanza/corenlp/client.py (hunk 1 above)
Comment (futurulus) [inline stanza/corenlp/client.py]: `raise NotImplementedError` and `raise NotImplementedError()` do the same thing (and I don't think PEP8 dictates a preference). If you want to make this an in-house style rule, I guess I wouldn't object, but it doesn't seem particularly necessary to me...

Not to be confused with the unrelated object called `NotImplemented`, used to indicate lack of operator overloading. `raise NotImplemented` is categorically wrong.

**Comment by futurulus** [comment]
Sweet, nice work @kelvinguu! I added the simplest possible implementation for a bunch of the Simple CoreNLP API methods--with those this module is already highly usable.

I like @vzhong 's suggestion of raising an exception if the field doesn't exist. My implementations don't check for that.

**Comment by kelvinguu** [comment]
Ok, new changes:
- The Sentence object is now a sequence of `Token`s rather than strings, as per @futurulus's suggestion
- Directories have been restructured
- you can convert the server's JSON response to AnnotatedDocument now. This is because the JSON actually contains some things which still aren't supported by the protobuf serializer.

We still need to implement @vzhong's suggestion about raising an exception if a field is missing. Right now, it just blindly returns the default value that the protobuf makes up...

**Comment by kelvinguu** [comment]
Merged into develop.

**Pull request closed without merging**